packages/wdio-browserstack-service/src/Percy/Percy-Handler.ts

```diff
@@ -0,0 +1,184 @@
+                this._percyScreenshotCounter -= 1
+            }
+        } catch (err: any) {
+            this._percyScreenshotCounter -= 1;
+            (this._browser.percyCaptureMap as PercyCaptureMap).decrement(sessionName ? sessionName : (this.sessionName as string), eventName as string)
+            PercyLogger.error(`Error while trying to auto capture Percy screenshot ${err}`)
+        }
+    }
+
+    async before () {
+        this._browser.percyCaptureMap = new PercyCaptureMap()
+    }
+
+    deferCapture(sessionName: string, eventName: string | null) {
+        /* Service doesn't wait for handling of browser commands so the below counter is used in teardown method to delay service exit */
+        this._percyScreenshotCounter += 1
+        this.percyDeferredScreenshots.push({ sessionName, eventName })
+    }
+
+    isDOMChangingCommand(args: BeforeCommandArgs): boolean {
+        /*
+          Percy screenshots which are to be taken on events such as send keys, element click & screenshot are deferred until
+          another DOM changing command is seen such that any DOM processing post the previous command is completed
+        */
+        return (
+            typeof args.method === 'string' && typeof args.endpoint === 'string' &&
+            (
+                (
+                    args.method === 'POST' &&
+                    (
+                        PERCY_DOM_CHANGING_COMMANDS_ENDPOINTS.includes(args.endpoint) ||
+                        (
+                            /* click / clear element */
+                            args.endpoint.includes('/session/:sessionId/element') &&
+                            (
+                                args.endpoint.includes('click') ||
+                                args.endpoint.includes('clear')
+                            )
+                        ) ||
+                        /* execute script sync / async */
+                        (args.endpoint.includes('/session/:sessionId/execute') && args.body?.script) ||
+                        /* Touch action for Appium */
+                        (args.endpoint.includes('/session/:sessionId/touch'))
+                    )
+                ) ||
+                ( args.method === 'DELETE' && args.endpoint === '/session/:sessionId' )
+            )
+        )
+    }
+
+    async cleanupDeferredScreenshots() {
+        this.isPercyCleanupProcessingUnderway = true
+        for (const entry of this.percyDeferredScreenshots) {
+            await this.percyAutoCapture(entry.eventName, entry.sessionName)
+        }
+        this.percyDeferredScreenshots = []
+        this.isPercyCleanupProcessingUnderway = false
+    }
+
+    async browserBeforeCommand (args: BeforeCommandArgs) {
+        try {
+            if (!this.isDOMChangingCommand(args)) {
+                return
+            }
+            do {
+                await sleep(1000)
+            } while (this.percyScreenshotInterval)
```

Comment:
It doesn't seem like it makes sense to have a do/while loop here if `this.percyScreenshotInterval` will be assigned right after, so the while condition will always be falsy
```suggestion
```